datafusion/sqllogictest/test_files/map.slt

```diff
@@ -131,17 +131,23 @@ SELECT MAKE_MAP([1,2], ['a', 'b'], [3,4], ['b']);
 ----
 {[1, 2]: [a, b], [3, 4]: [b]}
 
-query error
+query ?
 SELECT MAKE_MAP('POST', 41, 'HEAD', 'ab', 'PATCH', 30);
+----
+{POST: 41, HEAD: ab, PATCH: 30}
```

Comment:
I expected the query would fail because similar behavior isn't allowed in other databases (e.g. DuckDB). However, seems `make_array` will coercion the value to find a suitable type for them. In this case, all of them will be converted to `utf8`.
```

> select make_array(1,'a',3);
+-----------------------------------------+
| make_array(Int64(1),Utf8("a"),Int64(3)) |
+-----------------------------------------+
| [1, a, 3]                               |
+-----------------------------------------+
1 row(s) fetched. 
Elapsed 0.004 seconds.

> select arrow_typeof(make_array(1,'a',3));
+-----------------------------------------------------------------------------------------------------------------+
| arrow_typeof(make_array(Int64(1),Utf8("a"),Int64(3)))                                                           |
+-----------------------------------------------------------------------------------------------------------------+
| List(Field { name: "item", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }) |
+-----------------------------------------------------------------------------------------------------------------+
1 row(s) fetched. 
Elapsed 0.002 seconds.
```
I think if DataFusion allows this type of coercion for `make_array`, we can allow it for `make_map` too.

Reply:
I think we need another `make_array` to does not apply coercion. I prefer to align the behaviour to other system unless there is a good reason not to.

Reply:
I see. Maybe I can create a scalar function `make_array_strict` that won't implement the `coerce_types` method for `ScalarUDFImpl`, but other implementations are the same as `make_array`.
WDYT?

Reply:
> I think we need another `make_array` to does not apply coercion. I prefer to align the behaviour to other system unless there is a good reason not to.

Instead can we pass a boolean arg `should_coercion` with default value as false, to control such behaviour

Reply:
> > I think we need another `make_array` to does not apply coercion. I prefer to align the behaviour to other system unless there is a good reason not to.
> 
> Instead can we pass a boolean arg `should_coercion` with default value as false, to control such behaviour

The coercion logic is not simply work like if-else statement. The `make_array_inner` doesn't care about coercion, the coercion is in `type_coercion` pass in analzyer.

Reply:
Agreed. That's why I planned to implement another scalar function for it.

Reply:
> I think we need another `make_array` to does not apply coercion. I prefer to align the behaviour to other system unless there is a good reason not to.

I did more tests for DuckDB behavior and found something interesting. I found they also try to coercion types when building arrays or maps.
I arranged some notes for the behaviors:

## How DuckDB build a map
It seems that they also transform to two lists and call map function. Just like my first design, using `make_array`.
```sql
D select map {1:102, 2:20};
┌───────────────────────────────────────────────────────────┐
│ main.map(main.list_value(1, 2), main.list_value(102, 20)) │
│                   map(integer, integer)                   │
├───────────────────────────────────────────────────────────┤
│ {1=102, 2=20}                                             │
└───────────────────────────────────────────────────────────┘
```

## How DuckDB and DataFusion coercion array type
### DuckDB
- Array constructed from `INT32 and numeric string`: DuckDB will make it be `INTEGER[]`.
```sql
D select array[1,2,'3'];
┌────────────────────┐
│ (ARRAY[1, 2, '3']) │
│      int32[]       │
├────────────────────┤
│ [1, 2, 3]          │
└────────────────────┘
D select typeof(array[1,2,'3']);
┌────────────────────────────┐
│ typeof((ARRAY[1, 2, '3'])) │
│          varchar           │
├────────────────────────────┤
│ INTEGER[]                  │
└────────────────────────────┘
```
- Array constructed from `INT32 and non-numeric string`: DuckDB can't construct the array.
```sql
D select array[1,2,'a'];
Conversion Error: Could not convert the string 'a' to INT32
LINE 1: select array[1,2,'a'];
```
### DataFusion
- Array constructed from `INT32 and numeric string`: DataFusion will make it be `Uf8 array`.
```sql
> select [1,2,'1'];
+-----------------------------------------+
| make_array(Int64(1),Int64(2),Utf8("1")) |
+-----------------------------------------+
| [1, 2, 1]                               |
+-----------------------------------------+
1 row(s) fetched. 
Elapsed 0.001 seconds.
  
> select arrow_typeof([1,2,'1']);
+-----------------------------------------------------------------------------------------------------------------+
| arrow_typeof(make_array(Int64(1),Int64(2),Utf8("1")))                                                           |
+-----------------------------------------------------------------------------------------------------------------+
| List(Field { name: "item", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }) |
+-----------------------------------------------------------------------------------------------------------------+
1 row(s) fetched. 
Elapsed 0.001 seconds.
```
- Array constructed from `INT32 and non-numeric string`: DataFusion will make it be `Uf8 array`.
```sql
> select [1,2,'a'];
+-----------------------------------------+
| make_array(Int64(1),Int64(2),Utf8("a")) |
+-----------------------------------------+
| [1, 2, a]                               |
+-----------------------------------------+
1 row(s) fetched. 
Elapsed 0.001 seconds.

> select arrow_typeof([1,2,'a']);
+-----------------------------------------------------------------------------------------------------------------+
| arrow_typeof(make_array(Int64(1),Int64(2),Utf8("a")))                                                           |
+-----------------------------------------------------------------------------------------------------------------+
| List(Field { name: "item", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }) |
+-----------------------------------------------------------------------------------------------------------------+
1 row(s) fetched. 
Elapsed 0.001 seconds.
```
The behavior of type coercion between INT32 and String is really different. 

## How DuckDB coercion map type
- INT32 value and numeric string value: We can find the value `'20'` has been converted to `20`.
```sql
D select map {1:10, 2:'20'};
┌────────────────────────────────────────────────────────────┐
│ main.map(main.list_value(1, 2), main.list_value(10, '20')) │
│                   map(integer, integer)                    │
├────────────────────────────────────────────────────────────┤
│ {1=10, 2=20}                                               │
└────────────────────────────────────────────────────────────┘
```
- INT32 value and non-numeric string value. (It's what I tried in the first time. That's why I thought it shouldn't be allowed)
```sql
D select map {1:10, 2:'abc'};
Conversion Error: Could not convert string 'abc' to INT32
LINE 1: select map {1:10, 2:'abc'};
                            ^
```

## Conclusion
Referring to these behaviors, I think we can just back to using `make_array` to implement this. Because the behavior of type coercion is different, Our `make_map` can allow `map {1:10, 2:'a'}` but DuckDB can't do it. It makes sense for me.
@jayzhan211  WDYT?

Reply:
Alright, so the behaviour is actually depend on `array` itself.
I think we can use `make_array` for this.

But, if we want to introduce nice dataframe API `map(keys: Vec<Expr>, values: Vec<Expr>)`, I think we still need to pass `Vec<Expr>` instead of the result of `make_array`. However, we can introduce that in another PR.

Reply:
> Alright, so the behaviour is actually depend on `array` itself. I think we can use `make_array` in this case.

Ok, I'll roll back to `make_array` first.

> current API expects `map(vec![make_array(vec![lit("a"), lit("b")]), make_array(vec![lit("1"), lit("2")])])` a slightly better API is `map(vec![lit("a"), lit("b")], vec![lit(1), lit(2)])`

I'm not very familiar with the data frame implementation. Curiously, does the API for data frames also use the UDF `map`? I think the UDF is a logical layer function, but we don't have a corresponding logical expression for `vec!` other than `make_array`.

Reply:
dataframe API is used for building `Expr`. 

`map(vec![make_array(vec![lit("a"), lit("b")]), make_array(vec![lit("1"), lit("2")])])` is actually like `Expr::ScalarFunction(map_udf(), args: ...)`.

The idea is something like
```rust
fn map(keys: Vec<Expr>, values: Vec<Expr>) {
    let args: Vec<Expr> =  concat (keys, values)
    Expr::ScalarFunction(map_udf(), args)
}
```